polus-color-pyramid-builder-plugin/src/main.py

```diff
@@ -0,0 +1,201 @@
+        
+        # Resort the pixels
+        upper_bound_vals = np.sort(upper_bound_vals,axis=None)
+        lower_bound_vals = np.sort(lower_bound_vals,axis=None)
+    
+    return [lower_bound_vals[lower_bound_size],upper_bound_vals[-upper_bound_size]]
+
+if __name__=="__main__":
+    # Initialize the logger
+    logging.basicConfig(format='%(asctime)s - %(name)-8s - %(levelname)-8s - %(message)s',
+                        datefmt='%d-%b-%y %H:%M:%S')
+    logger = logging.getLogger("main")
+    logger.setLevel(logging.INFO)
+
+    ''' Argument parsing '''
+    logger.info("Parsing arguments...")
+    parser = argparse.ArgumentParser(prog='main', description='Builds a DeepZoom color pyramid.')
+    
+    # Input arguments
+    parser.add_argument('--filePattern', dest='filePattern', type=str,
+                        help='Filename pattern used to separate data', required=True)
+    parser.add_argument('--inpDir', dest='inpDir', type=str,
+                        help='Input image collection to be processed by this plugin', required=True)
+    parser.add_argument('--layout', dest='layout', type=str,
+                        help='Color ordering (e.g. 1,11,,,,5,6)', required=True)
+    parser.add_argument('--bounds', dest='bounds', type=str,
+                        help='Set bounds (should be float-float, int-int, or blank, e.g. 0.01-0.99,0-16000,,,,,)', required=False)
+    
+    # Output arguments
+    parser.add_argument('--outDir', dest='outDir', type=str,
+                        help='Output pyramid path.', required=True)
+    
+    # Parse the arguments
+    args = parser.parse_args()
+    filePattern = args.filePattern
+    logger.info('filePattern = {}'.format(filePattern))
+    inpDir = args.inpDir
+    if (Path.is_dir(Path(args.inpDir).joinpath('images'))):
+        # switch to images folder if present
+        fpath = str(Path(args.inpDir).joinpath('images').absolute())
+    logger.info('inpDir = {}'.format(inpDir))
+    layout = args.layout
+    logger.info('layout = {}'.format(layout))
+    bounds = args.bounds
+    logger.info('bounds = {}'.format(bounds))
+    outDir = args.outDir
+    logger.info('outDir = {}'.format(outDir))
+    outDir = Path(outDir)
+    
+    # Parse the layout
+    layout = [None if l=='' else int(l) for l in layout.split(',')]
+    if len(layout)>7:
+        layout = layout[:7]
+        
+    # Parse the bounds
+    if bounds != None:
+        bounds = [[None] if l=='' else get_number(l) for l in bounds.split(',')]
+        bounds = bounds[:len(layout)]
+    else:
+        bounds = [[None] for _ in layout]
+        
+    # Parse files
+    fp = filepattern.FilePattern(inpDir,filePattern)
+    
+    count = 0
+    for files in fp.iterate(group_by='c'):
+        outDirFrame = outDir.joinpath('{}_files'.format(count))
+        outDirFrame.mkdir()
+        count += 1
+        bioreaders = []
+        threads = []
+        with ThreadPoolExecutor(max([multiprocessing.cpu_count()//2,2])) as executor:
```

Comment:
Since this uses bfio with the Python backend, the read and writing are already multithreaded. This means we are starting threads that have multithreaded operations. I could have set this to something else, but setting it to use threads=cpu_count would use more memory. This seemed like a good compromise. I haven't extensively tested.